File: sysbridge/src/main/cpp/android/input/KeyLayoutMap.cpp
/*
 * Copyright (C) 2008 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, Tokenizer.cppeither express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../../logging.h"
#include <android/keycodes.h>
#include "../utils/String8.h"
#include "KeyLayoutMap.h"
#include "../utils/Tokenizer.h"
#include "InputEventLabels.h"
#include <unordered_map>

#include <cstdlib>
#include <string_view>
#include <unordered_map>
#include "../libbase/result.h"
#include "../liblog/log_main.h"
#include "Input.h"

#define DEBUG_MAPPING false
#define DEBUG_PARSER false

// Enables debug output for parser performance.
#define DEBUG_PARSER_PERFORMANCE 0

namespace android {

    namespace {

        std::optional<int> parseInt(const char *str) {
            char *end;
            errno = 0;
            const int value = strtol(str, &end, 0);
            if (end == str) {
                LOGE("Could not parse %s", str);
                return {};
            }
            if (errno == ERANGE) {
                LOGE("Out of bounds: %s", str);
                return {};
            }
            return value;
        }

        constexpr const char *WHITESPACE = " \t\r";

    } // namespace

    KeyLayoutMap::KeyLayoutMap() = default;

    KeyLayoutMap::~KeyLayoutMap() = default;

    base::Result<std::shared_ptr<KeyLayoutMap>>
    KeyLayoutMap::loadContents(const std::string &filename,
                               const char *contents) {
        return load(filename, contents);
    }

    base::Result<std::shared_ptr<KeyLayoutMap>> KeyLayoutMap::load(const std::string &filename,
                                                                   const char *contents) {
        Tokenizer *tokenizer;
        status_t status;
        if (contents == nullptr) {
            status = Tokenizer::open(String8(filename.c_str()), &tokenizer);
        } else {
            status = Tokenizer::fromContents(String8(filename.c_str()), contents, &tokenizer);
        }

        if (status) {
            LOGE("Error %d opening key layout map file %s.", status, filename.c_str());
            return Errorf("Error {} opening key layout map file {}.", status, filename.c_str());
        }
        std::unique_ptr<Tokenizer> t(tokenizer);
        auto ret = load(t.get());
        if (!ret.ok()) {
            return ret;
        }
        const std::shared_ptr<KeyLayoutMap> &map = *ret;
        LOG_ALWAYS_FATAL_IF(map == nullptr, "Returned map should not be null if there's no error");

        map->mLoadFileName = filename;
        return ret;
    }

    base::Result<std::shared_ptr<KeyLayoutMap>> KeyLayoutMap::load(Tokenizer *tokenizer) {
        std::shared_ptr<KeyLayoutMap> map = std::shared_ptr<KeyLayoutMap>(new KeyLayoutMap());
        status_t status = OK;
        if (!map.get()) {
            LOGE("Error allocating key layout map.");
            return Errorf("Error allocating key layout map.");
        } else {
#if DEBUG_PARSER_PERFORMANCE
            nsecs_t startTime = systemTime(SYSTEM_TIME_MONOTONIC);
#endif
            Parser parser(map.get(), tokenizer);
            status = parser.parse();
#if DEBUG_PARSER_PERFORMANCE
            nsecs_t elapsedTime = systemTime(SYSTEM_TIME_MONOTONIC) - startTime;
            LOGD("Parsed key layout map file '%s' %d lines in %0.3fms.",
                  tokenizer->getFilename().c_str(), tokenizer->getLineNumber(),
                  elapsedTime / 1000000.0);
#endif
            LOGE("PARSE STATUS = %d", status);
            if (!status) {
                return std::move(map);
            }
        }
        return Errorf("Load KeyLayoutMap failed {}.", status);
    }

    status_t KeyLayoutMap::mapKey(int32_t scanCode, int32_t usageCode,
                                  int32_t *outKeyCode, uint32_t *outFlags) const {
        const Key *key = getKey(scanCode, usageCode);
        if (!key) {
            ALOGD_IF(DEBUG_MAPPING, "mapKey: scanCode=%d, usageCode=0x%08x ~ Failed.", scanCode,
                     usageCode);
            *outKeyCode = AKEYCODE_UNKNOWN;
            *outFlags = 0;
            return NAME_NOT_FOUND;
        }

        *outKeyCode = key->keyCode;
        *outFlags = key->flags;

        ALOGD_IF(DEBUG_MAPPING,
                 "mapKey: scanCode=%d, usageCode=0x%08x ~ Result keyCode=%d, outFlags=0x%08x.",
                 scanCode, usageCode, *outKeyCode, *outFlags);
        return NO_ERROR;
    }

    const KeyLayoutMap::Key *KeyLayoutMap::getKey(int32_t scanCode, int32_t usageCode) const {
        if (usageCode) {
            auto it = mKeysByUsageCode.find(usageCode);
            if (it != mKeysByUsageCode.end()) {
                return &it->second;
            }
        }
        if (scanCode) {
            auto it = mKeysByScanCode.find(scanCode);
            if (it != mKeysByScanCode.end()) {
                return &it->second;
            }
        }
        return nullptr;
    }

    std::vector<int32_t> KeyLayoutMap::findScanCodesForKey(int32_t keyCode) const {
        std::vector<int32_t> scanCodes;
        // b/354333072: Only consider keys without FUNCTION flag
        for (const auto &[scanCode, key]: mKeysByScanCode) {
            if (keyCode == key.keyCode && !(key.flags & POLICY_FLAG_FUNCTION)) {
                scanCodes.push_back(scanCode);
            }
        }
        return scanCodes;
    }

    std::vector<int32_t> KeyLayoutMap::findUsageCodesForKey(int32_t keyCode) const {
        std::vector<int32_t> usageCodes;
        for (const auto &[usageCode, key]: mKeysByUsageCode) {
            if (keyCode == key.keyCode && !(key.flags & POLICY_FLAG_FALLBACK_USAGE_MAPPING)) {
                usageCodes.push_back(usageCode);
            }
        }
        return usageCodes;
    }

    std::optional<AxisInfo> KeyLayoutMap::mapAxis(int32_t scanCode) const {
        auto it = mAxes.find(scanCode);
        if (it == mAxes.end()) {
            ALOGD_IF(DEBUG_MAPPING, "mapAxis: scanCode=%d ~ Failed.", scanCode);
            return std::nullopt;
        }

        const AxisInfo &axisInfo = it->second;
        ALOGD_IF(DEBUG_MAPPING,
                 "mapAxis: scanCode=%d ~ Result mode=%d, axis=%d, highAxis=%d, "
                 "splitValue=%d, flatOverride=%d.",
                 scanCode, axisInfo.mode, axisInfo.axis, axisInfo.highAxis, axisInfo.splitValue,
                 axisInfo.flatOverride);
        return axisInfo;
    }

// --- KeyLayoutMap::Parser ---

    KeyLayoutMap::Parser::Parser(KeyLayoutMap *map, Tokenizer *tokenizer) :
            mMap(map), mTokenizer(tokenizer) {
    }

    KeyLayoutMap::Parser::~Parser() {
    }

    status_t KeyLayoutMap::Parser::parse() {
        while (!mTokenizer->isEof()) {
            ALOGD_IF(DEBUG_PARSER, "Parsing %s: '%s'.", mTokenizer->getLocation().c_str(),
                     mTokenizer->peekRemainderOfLine().c_str());

            mTokenizer->skipDelimiters(WHITESPACE);

            if (!mTokenizer->isEol() && mTokenizer->peekChar() != '#') {
                String8 keywordToken = mTokenizer->nextToken(WHITESPACE);
                if (keywordToken == "key") {
                    mTokenizer->skipDelimiters(WHITESPACE);
                    status_t status = parseKey();
                    if (status) return status;
                } else if (keywordToken == "axis") {
                    mTokenizer->skipDelimiters(WHITESPACE);
                    status_t status = parseAxis();
                    if (status) return status;
                } else if (keywordToken == "led") {
                    // Skip LEDs, we don't need them for Key Mapper
                    mTokenizer->nextLine();
                    continue;
                } else if (keywordToken == "sensor") {
                    // Skip Sensors, we don't need them for Key Mapper
                    mTokenizer->nextLine();
                    continue;
                } else if (keywordToken == "requires_kernel_config") {
                    mTokenizer->skipDelimiters(WHITESPACE);
                    status_t status = parseRequiredKernelConfig();
                    if (status) return status;
                } else {
                    LOGE("%s: Expected keyword, got '%s'.", mTokenizer->getLocation().c_str(),
                         keywordToken.c_str());
                    return BAD_VALUE;
                }

                mTokenizer->skipDelimiters(WHITESPACE);

                if (!mTokenizer->isEol() && mTokenizer->peekChar() != '#') {
                    LOGW("%s: Expected end of line or trailing comment, got '%s'.",
                         mTokenizer->getLocation().c_str(),
                         mTokenizer->peekRemainderOfLine().c_str());
                    return BAD_VALUE;
                }
            }

            mTokenizer->nextLine();
        }
        return NO_ERROR;
    }

    status_t KeyLayoutMap::Parser::parseKey() {
        String8 codeToken = mTokenizer->nextToken(WHITESPACE);
        bool mapUsage = false;
        if (codeToken == "usage") {
            mapUsage = true;
            mTokenizer->skipDelimiters(WHITESPACE);
            codeToken = mTokenizer->nextToken(WHITESPACE);
        }

        std::optional<int> code = parseInt(codeToken.c_str());
        if (!code) {
            LOGE("%s: Expected key %s number, got '%s'.", mTokenizer->getLocation().c_str(),
                 mapUsage ? "usage" : "scan code", codeToken.c_str());
            return BAD_VALUE;
        }
        std::unordered_map<int32_t, Key> &map =
                mapUsage ? mMap->mKeysByUsageCode : mMap->mKeysByScanCode;
        if (map.find(*code) != map.end()) {
            LOGE("%s: Duplicate entry for key %s '%s'.", mTokenizer->getLocation().c_str(),
                 mapUsage ? "usage" : "scan code", codeToken.c_str());
            return BAD_VALUE;
        }

        mTokenizer->skipDelimiters(WHITESPACE);
        String8 keyCodeToken = mTokenizer->nextToken(WHITESPACE);
        std::optional<int> keyCode = InputEventLookup::getKeyCodeByLabel(keyCodeToken.c_str());

        if (!keyCode) {
            LOGW("%s: Unknown key code label %s", mTokenizer->getLocation().c_str(),
                 keyCodeToken.c_str());
            // Do not crash at this point because there may be more flags afterwards that need parsing.
        }

        uint32_t flags = 0;
        for (;;) {
            mTokenizer->skipDelimiters(WHITESPACE);
            if (mTokenizer->isEol() || mTokenizer->peekChar() == '#') break;

            String8 flagToken = mTokenizer->nextToken(WHITESPACE);
            std::optional<int> flag = InputEventLookup::getKeyFlagByLabel(flagToken.c_str());
            if (!flag) {
                LOGE("%s: Expected key flag label, got '%s'.", mTokenizer->getLocation().c_str(),
                     flagToken.c_str());
                return BAD_VALUE;
            }
            if (flags & *flag) {
                LOGE("%s: Duplicate key flag '%s'.", mTokenizer->getLocation().c_str(),
                     flagToken.c_str());
                return BAD_VALUE;
            }
            flags |= *flag;
        }

        ALOGD_IF(DEBUG_PARSER, "Parsed key %s: code=%d, keyCode=%d, flags=0x%08x.",
                 mapUsage ? "usage" : "scan code", *code, *keyCode, flags);

        // The key code may be unknown so only insert a key if it is known.
        if (keyCode) {
            Key key;
            key.keyCode = *keyCode;
            key.flags = flags;
            map.insert({*code, key});
        }

        return NO_ERROR;
    }

    status_t KeyLayoutMap::Parser::parseAxis() {
        String8 scanCodeToken = mTokenizer->nextToken(WHITESPACE);
        std::optional<int> scanCode = parseInt(scanCodeToken.c_str());
        if (!scanCode) {
            LOGE("%s: Expected axis scan code number, got '%s'.", mTokenizer->getLocation().c_str(),
                 scanCodeToken.c_str());
            return BAD_VALUE;
        }
        if (mMap->mAxes.find(*scanCode) != mMap->mAxes.end()) {
            LOGE("%s: Duplicate entry for axis scan code '%s'.", mTokenizer->getLocation().c_str(),
                 scanCodeToken.c_str());
            return BAD_VALUE;
        }

        AxisInfo axisInfo;

        mTokenizer->skipDelimiters(WHITESPACE);
        String8 token = mTokenizer->nextToken(WHITESPACE);
        if (token == "invert") {
            axisInfo.mode = AxisInfo::MODE_INVERT;

            mTokenizer->skipDelimiters(WHITESPACE);
            String8 axisToken = mTokenizer->nextToken(WHITESPACE);
            std::optional<int> axis = InputEventLookup::getAxisByLabel(axisToken.c_str());
            if (!axis) {
                LOGE("%s: Expected inverted axis label, got '%s'.",
                     mTokenizer->getLocation().c_str(),
                     axisToken.c_str());
                return BAD_VALUE;
            }
            axisInfo.axis = *axis;
        } else if (token == "split") {
            axisInfo.mode = AxisInfo::MODE_SPLIT;

            mTokenizer->skipDelimiters(WHITESPACE);
            String8 splitToken = mTokenizer->nextToken(WHITESPACE);
            std::optional<int> splitValue = parseInt(splitToken.c_str());
            if (!splitValue) {
                LOGE("%s: Expected split value, got '%s'.", mTokenizer->getLocation().c_str(),
                     splitToken.c_str());
                return BAD_VALUE;
            }
            axisInfo.splitValue = *splitValue;

            mTokenizer->skipDelimiters(WHITESPACE);
            String8 lowAxisToken = mTokenizer->nextToken(WHITESPACE);
            std::optional<int> axis = InputEventLookup::getAxisByLabel(lowAxisToken.c_str());
            if (!axis) {
                LOGE("%s: Expected low axis label, got '%s'.", mTokenizer->getLocation().c_str(),
                     lowAxisToken.c_str());
                return BAD_VALUE;
            }
            axisInfo.axis = *axis;

            mTokenizer->skipDelimiters(WHITESPACE);
            String8 highAxisToken = mTokenizer->nextToken(WHITESPACE);
            std::optional<int> highAxis = InputEventLookup::getAxisByLabel(highAxisToken.c_str());
            if (!highAxis) {
                LOGE("%s: Expected high axis label, got '%s'.", mTokenizer->getLocation().c_str(),
                     highAxisToken.c_str());
                return BAD_VALUE;
            }
            axisInfo.highAxis = *highAxis;
        } else {
            std::optional<int> axis = InputEventLookup::getAxisByLabel(token.c_str());
            if (!axis) {
                LOGE("%s: Expected axis label, 'split' or 'invert', got '%s'.",
                     mTokenizer->getLocation().c_str(), token.c_str());
                return BAD_VALUE;
            }
            axisInfo.axis = *axis;
        }

        for (;;) {
            mTokenizer->skipDelimiters(WHITESPACE);
            if (mTokenizer->isEol() || mTokenizer->peekChar() == '#') {
                break;
            }
            String8 keywordToken = mTokenizer->nextToken(WHITESPACE);
            if (keywordToken == "flat") {
                mTokenizer->skipDelimiters(WHITESPACE);
                String8 flatToken = mTokenizer->nextToken(WHITESPACE);
                std::optional<int> flatOverride = parseInt(flatToken.c_str());
                if (!flatOverride) {
                    LOGE("%s: Expected flat value, got '%s'.", mTokenizer->getLocation().c_str(),
                         flatToken.c_str());
                    return BAD_VALUE;
                }
                axisInfo.flatOverride = *flatOverride;
            } else {
                LOGE("%s: Expected keyword 'flat', got '%s'.", mTokenizer->getLocation().c_str(),
                     keywordToken.c_str());
                return BAD_VALUE;
            }
        }

        ALOGD_IF(DEBUG_PARSER,
                 "Parsed axis: scanCode=%d, mode=%d, axis=%d, highAxis=%d, "
                 "splitValue=%d, flatOverride=%d.",
                 *scanCode, axisInfo.mode, axisInfo.axis, axisInfo.highAxis, axisInfo.splitValue,
                 axisInfo.flatOverride);
        mMap->mAxes.insert({*scanCode, axisInfo});
        return NO_ERROR;
    }

// Parse the name of a required kernel config.
// The layout won't be used if the specified kernel config is not present
// Examples:
// requires_kernel_config CONFIG_HID_PLAYSTATION
    status_t KeyLayoutMap::Parser::parseRequiredKernelConfig() {
        String8 codeToken = mTokenizer->nextToken(WHITESPACE);
        std::string configName = codeToken.c_str();

        const auto result = mMap->mRequiredKernelConfigs.emplace(configName);
        if (!result.second) {
            LOGE("%s: Duplicate entry for required kernel config %s.",
                 mTokenizer->getLocation().c_str(), configName.c_str());
            return BAD_VALUE;
        }

//        ALOGD_IF(DEBUG_PARSER, "Parsed required kernel config: name=%s", configName.c_str());
        return NO_ERROR;
    }
} // namespace android

File: evdev/src/main/rust/evdev_manager/src/key_layout_map.rs
//! Key layout map parser for Android key layout files.
//!
//! This module provides functionality to parse Android key layout (.kl) files
//! and map scan codes to Android key codes.
//!
//! AOSP keylayout files can be found at:
//! https://cs.android.com/android/platform/superproject/+/android-latest-release:frameworks/base/data/keyboards/

use crate::android::android_codes::POLICY_FLAG_FUNCTION;
use crate::input_event_lookup::{get_axis_by_label, get_key_code_by_label, get_key_flag_by_label};
use crate::tokenizer::Tokenizer;
use std::collections::HashMap;
use std::fmt::format;

/// Describes a mapping from keyboard scan codes to Android key codes.
///
/// This object is immutable after it has been loaded.
pub struct KeyLayoutMap {
    keys_by_scan_code: HashMap<u32, KeyLayoutKey>,
    axes: HashMap<u32, KeyLayoutAxisInfo>,
}

/// Represents a key mapping entry.
#[derive(Debug, Clone)]
pub struct KeyLayoutKey {
    pub key_code: u32,
    flags: u32,
}

/// Represents axis information for joystick/gamepad axes.
#[derive(Debug, Clone)]
pub struct KeyLayoutAxisInfo {
    mode: KeyLayoutAxisMode,
    axis: u32,
    high_axis: Option<u32>,
    split_value: Option<i32>,
    flat_override: Option<i32>,
}

/// Axis mapping mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum KeyLayoutAxisMode {
    /// Axis value is reported directly.
    Normal,
    /// Axis value should be inverted before reporting.
    Invert,
    /// Axis value should be split into two axes.
    Split,
}

impl KeyLayoutMap {
    /// Load a key layout map from a file path.
    pub fn load_from_file(file_path: &str) -> Result<Self, String> {
        let tokenizer = Tokenizer::from_file(file_path)?;
        Self::load(tokenizer)
    }

    /// Load a key layout map from file contents (useful for testing).
    pub fn load_from_contents(contents: &str) -> Result<Self, String> {
        let tokenizer = Tokenizer::from_contents("", contents);
        Self::load(tokenizer)
    }

    fn load(mut tokenizer: Tokenizer) -> Result<Self, String> {
        let mut map = KeyLayoutMap {
            keys_by_scan_code: HashMap::new(),
            axes: HashMap::new(),
        };

        let mut parser = Parser::new(&mut map, &mut tokenizer);
        parser.parse()?;

        Ok(map)
    }

    /// Map a scan code to an Android key code.
    ///
    /// Returns `Ok((key_code, flags))` on success, or
    pub fn map_key(&self, scan_code: u32) -> Option<KeyLayoutKey> {
        self.keys_by_scan_code.get(&scan_code).cloned()
    }

    /// Map a scan code to axis information.
    ///
    /// Returns `Some(axis_info)` if the scan code maps to an axis, or `None` if not found.
    pub fn map_axis(&self, scan_code: u32) -> Option<KeyLayoutAxisInfo> {
        self.axes.get(&scan_code).cloned()
    }

    /// Find all scan codes that map to the given key code.
    ///
    /// Only considers keys without the FUNCTION flag.
    pub fn find_scan_codes_for_key(&self, key_code: u32) -> Vec<u32> {
        self.keys_by_scan_code
            .iter()
            .filter_map(|(scan_code, key)| {
                if key.key_code == key_code && (key.flags & POLICY_FLAG_FUNCTION) == 0 {
                    Some(*scan_code)
                } else {
                    None
                }
            })
            .collect()
    }
}

/// Parser for key layout map files.
struct Parser<'a> {
    map: &'a mut KeyLayoutMap,
    tokenizer: &'a mut Tokenizer,
}

impl<'a> Parser<'a> {
    fn new(map: &'a mut KeyLayoutMap, tokenizer: &'a mut Tokenizer) -> Self {
        Self { map, tokenizer }
    }

    fn parse(&mut self) -> Result<(), String> {
        const WHITESPACE: &str = " \t\r";

        while !self.tokenizer.is_eof() {
            self.tokenizer.skip_delimiters(WHITESPACE);

            if !self.tokenizer.is_eol() && self.tokenizer.peek_char() != '#' {
                let keyword_token = self.tokenizer.next_token(WHITESPACE);
                match keyword_token.as_str() {
                    "key" => {
                        self.tokenizer.skip_delimiters(WHITESPACE);
                        self.parse_key()?;
                    }
                    "axis" => {
                        self.tokenizer.skip_delimiters(WHITESPACE);
                        self.parse_axis()?;
                    }
                    "led" | "sensor" => {
                        // Skip LEDs and sensors, we don't need them
                        self.tokenizer.next_line();
                        continue;
                    }
                    "usage" => {
                        // Skip usage code entries - evdev only provides scan codes
                        self.tokenizer.next_line();
                        continue;
                    }
                    _ => {
                        return Err(format!(
                            "{}: Expected keyword, got '{}'.",
                            self.tokenizer.get_location(),
                            keyword_token
                        ));
                    }
                }

                self.tokenizer.skip_delimiters(WHITESPACE);

                if !self.tokenizer.is_eol() && self.tokenizer.peek_char() != '#' {
                    return Err(format!(
                        "{}: Expected end of line or trailing comment, got '{}'.",
                        self.tokenizer.get_location(),
                        self.tokenizer.peek_remainder_of_line()
                    ));
                }
            }

            self.tokenizer.next_line();
        }

        Ok(())
    }

    fn parse_key(&mut self) -> Result<(), String> {
        const WHITESPACE: &str = " \t\r";

        let code_token = self.tokenizer.next_token(WHITESPACE);

        // Skip "usage" entries - we only support scan codes
        if code_token == "usage" {
            self.tokenizer.next_line();
            return Ok(());
        }

        let scan_code = parse_int(&code_token).ok_or_else(|| {
            format!(
                "{}: Expected key scan code number, got '{}'.",
                self.tokenizer.get_location(),
                code_token
            )
        })?;

        if scan_code < 0 {
            return Err(format!(
                "{} is not a valid key scan code. Negative numbers are not allowed.",
                scan_code
            ));
        }

        if self.map.keys_by_scan_code.contains_key(&(scan_code as u32)) {
            return Err(format!(
                "{}: Duplicate entry for key scan code '{}'.",
                self.tokenizer.get_location(),
                code_token
            ));
        }

        self.tokenizer.skip_delimiters(WHITESPACE);
        let key_code_token = self.tokenizer.next_token(WHITESPACE);
        let key_code = get_key_code_by_label(&key_code_token);

        let mut flags = 0u32;
        loop {
            self.tokenizer.skip_delimiters(WHITESPACE);
            if self.tokenizer.is_eol() || self.tokenizer.peek_char() == '#' {
                break;
            }

            let flag_token = self.tokenizer.next_token(WHITESPACE);
            let flag = get_key_flag_by_label(&flag_token).ok_or_else(|| {
                format!(
                    "{}: Expected key flag label, got '{}'.",
                    self.tokenizer.get_location(),
                    flag_token
                )
            })?;

            if (flags & flag) != 0 {
                return Err(format!(
                    "{}: Duplicate key flag '{}'.",
                    self.tokenizer.get_location(),
                    flag_token
                ));
            }
            flags |= flag;
        }

        // Only insert if the key code is known
        if let Some(key_code) = key_code {
            let key = KeyLayoutKey { key_code, flags };
            self.map.keys_by_scan_code.insert(scan_code as u32, key);
        }

        Ok(())
    }

    fn parse_axis(&mut self) -> Result<(), String> {
        const WHITESPACE: &str = " \t\r";

        let scan_code_token = self.tokenizer.next_token(WHITESPACE);
        let scan_code = parse_int(&scan_code_token).ok_or_else(|| {
            format!(
                "{}: Expected axis scan code number, got '{}'.",
                self.tokenizer.get_location(),
                scan_code_token
            )
        })?;

        if scan_code < 0 {
            return Err(format!(
                "{} is not a valid key scan code for an axis. Negative numbers are not allowed.",
                scan_code
            ));
        }

        if self.map.axes.contains_key(&(scan_code as u32)) {
            return Err(format!(
                "{}: Duplicate entry for axis scan code '{}'.",
                self.tokenizer.get_location(),
                scan_code_token
            ));
        }

        let mut axis_mode: KeyLayoutAxisMode = KeyLayoutAxisMode::Normal;
        let axis: u32;
        let mut split_value: Option<i32> = None;
        let mut high_axis: Option<u32> = None;
        let mut flat_override: Option<i32> = None;

        self.tokenizer.skip_delimiters(WHITESPACE);
        let token = self.tokenizer.next_token(WHITESPACE);

        if token == "invert" {
            axis_mode = KeyLayoutAxisMode::Invert;

            self.tokenizer.skip_delimiters(WHITESPACE);

            let axis_token = self.tokenizer.next_token(WHITESPACE);

            axis = get_axis_by_label(&axis_token).ok_or_else(|| {
                format!(
                    "{}: Expected inverted axis label, got '{}'.",
                    self.tokenizer.get_location(),
                    axis_token
                )
            })?;
        } else if token == "split" {
            axis_mode = KeyLayoutAxisMode::Split;

            self.tokenizer.skip_delimiters(WHITESPACE);
            let split_token = self.tokenizer.next_token(WHITESPACE);
            let split_value_raw = parse_int(&split_token).ok_or_else(|| {
                format!(
                    "{}: Expected split value, got '{}'.",
                    self.tokenizer.get_location(),
                    split_token
                )
            })?;
            split_value = Some(split_value_raw);

            self.tokenizer.skip_delimiters(WHITESPACE);
            let low_axis_token = self.tokenizer.next_token(WHITESPACE);
            axis = get_axis_by_label(&low_axis_token).ok_or_else(|| {
                format!(
                    "{}: Expected low axis label, got '{}'.",
                    self.tokenizer.get_location(),
                    low_axis_token
                )
            })?;

            self.tokenizer.skip_delimiters(WHITESPACE);
            let high_axis_token = self.tokenizer.next_token(WHITESPACE);
            let high_axis_raw = get_axis_by_label(&high_axis_token).ok_or_else(|| {
                format!(
                    "{}: Expected high axis label, got '{}'.",
                    self.tokenizer.get_location(),
                    high_axis_token
                )
            })?;
            high_axis = Some(high_axis_raw);
        } else {
            axis = get_axis_by_label(&token).ok_or_else(|| {
                format!(
                    "{}: Expected axis label, 'split' or 'invert', got '{}'.",
                    self.tokenizer.get_location(),
                    token
                )
            })?;
        }

        loop {
            self.tokenizer.skip_delimiters(WHITESPACE);
            if self.tokenizer.is_eol() || self.tokenizer.peek_char() == '#' {
                break;
            }

            let keyword_token = self.tokenizer.next_token(WHITESPACE);
            if keyword_token == "flat" {
                self.tokenizer.skip_delimiters(WHITESPACE);
                let flat_token = self.tokenizer.next_token(WHITESPACE);
                let flat_override_raw = parse_int(&flat_token).ok_or_else(|| {
                    format!(
                        "{}: Expected flat value, got '{}'.",
                        self.tokenizer.get_location(),
                        flat_token
                    )
                })?;
                flat_override = Some(flat_override_raw);
            } else {
                return Err(format!(
                    "{}: Expected keyword 'flat', got '{}'.",
                    self.tokenizer.get_location(),
                    keyword_token
                ));
            }
        }

        let axis_info: KeyLayoutAxisInfo = KeyLayoutAxisInfo {
            mode: axis_mode,
            axis,
            high_axis,
            split_value,
            flat_override,
        };

        self.map.axes.insert(scan_code as u32, axis_info);
        Ok(())
    }
}

/// Parse an integer from a string (supports decimal, hex with 0x prefix, and octal with 0 prefix).
fn parse_int(s: &str) -> Option<i32> {
    if s.is_empty() {
        return None;
    }

    // Handle hex (0x prefix) and octal (0 prefix)
    if s.starts_with("0x") || s.starts_with("0X") {
        i32::from_str_radix(&s[2..], 16).ok()
    } else if s.starts_with('0') && s.len() > 1 {
        i32::from_str_radix(&s[1..], 8).ok()
    } else {
        s.parse::<i32>().ok()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_int() {
        assert_eq!(parse_int("123"), Some(123));
        assert_eq!(parse_int("-999"), Some(-999));
        assert_eq!(parse_int("0x1a"), Some(26));
        assert_eq!(parse_int("0XFF"), Some(255));
        assert_eq!(parse_int("077"), Some(63)); // octal
        assert_eq!(parse_int(""), None);
        assert_eq!(parse_int("abc"), None);
    }

    #[test]
    fn test_load_from_contents_simple() {
        let contents = "key 1 ESCAPE\nkey 2 1\n";
        let map = KeyLayoutMap::load_from_contents(contents).unwrap();

        let key = map.map_key(1).unwrap();
        assert_eq!(key.key_code, 111); // ESCAPE
        assert_eq!(key.flags, 0);

        let key = map.map_key(2).unwrap();
        assert_eq!(key.key_code, 8); // KEYCODE_1
        assert_eq!(key.flags, 0);
    }

    #[test]
    fn test_load_from_contents_with_flags() {
        let contents = "key 465 ESCAPE FUNCTION\n";
        let map = KeyLayoutMap::load_from_contents(contents).unwrap();

        let key = map.map_key(465).unwrap();
        assert_eq!(key.key_code, 111); // ESCAPE
        assert_eq!(key.flags, 0x00000004); // FUNCTION flag
    }

    #[test]
    fn test_load_from_contents_axis() {
        let contents = "axis 0x00 X\naxis 0x01 Y\n";
        let map = KeyLayoutMap::load_from_contents(contents).unwrap();

        let axis_info = map.map_axis(0x00).unwrap();
        assert_eq!(axis_info.mode, KeyLayoutAxisMode::Normal);
        assert_eq!(axis_info.axis, 0); // X axis

        let axis_info = map.map_axis(0x01).unwrap();
        assert_eq!(axis_info.mode, KeyLayoutAxisMode::Normal);
        assert_eq!(axis_info.axis, 1); // Y axis
    }

    #[test]
    fn test_find_scan_codes_for_key() {
        let contents = "key 1 ESCAPE\nkey 465 ESCAPE FUNCTION\n";
        let map = KeyLayoutMap::load_from_contents(contents).unwrap();

        let scan_codes = map.find_scan_codes_for_key(111); // ESCAPE
        assert_eq!(scan_codes, vec![1]); // Only scan code 1, not 465 (has FUNCTION flag)
    }
}
